swarm/pss/pss.go

```diff
@@ -214,66 +162,86 @@ func (self *Pss) Register(topic *whisper.TopicType, handler Handler) func() {
+// Get a Public Key from id
+func (self *Pss) GetPeerPublicKey(pubkeyid string, topic whisper.TopicType) (*ecdsa.PublicKey, *pssPeer) {
+	pubkey := crypto.ToECDSAPub(common.FromHex(pubkeyid))
+	if pubkey.X == nil {
+		return nil, nil
+	}
+	return pubkey, self.pubKeyPool[pubkeyid][topic]
+}
+
+// Automatically generate a new symkey for a topic and address hint
+func (self *Pss) generateSymmetricKey(topic whisper.TopicType, address PssAddress, expires time.Duration, addtocache bool) (string, error) {
 	keyid, err := self.w.GenerateSymKey()
 	if err != nil {
 		return "", err
 	}
-	psp := self.findOrCreatePssPeer(addr, topic)
-	self.lock.Lock()
-	defer self.lock.Unlock()
-	if _, ok := self.peerPool[addr]; !ok {
-		return "", fmt.Errorf("no address entry %x in peerpool", addr)
-	}
-	psp.recvsymkey = keyid
-	psp.symkeyexpires = time.Now().Add(time.Hour * 24 * 365)
-	self.reverseSymKeyPool[keyid] = addr
-	self.symkeycachecursor++
-	self.symkeycache[self.symkeycachecursor%cap(self.symkeycache)] = psp
+	self.addSymmetricKeyToPool(keyid, topic, address, expires, addtocache)
 	return keyid, nil
 }
 
-// Set the symmetric key for outgoing communications
-// this is the key received when receiving an incoming handshake
-func (self *Pss) SetOutgoingSymmetricKey(addr pot.Address, topic whisper.TopicType, key []byte) (string, error) {
+// Manually set a new symkey for a topic and address hint
+//
+// If addtocache is set to true, the key will be added to the collection of keys used to attempt incoming message decryption
+func (self *Pss) SetSymmetricKey(key []byte, topic whisper.TopicType, address PssAddress, expires time.Duration, addtocache bool) (string, error) {
 	keyid, err := self.w.AddSymKeyDirect(key)
 	if err != nil {
 		return "", err
 	}
-	psp := self.findOrCreatePssPeer(addr, topic)
-	self.lock.Lock()
-	defer self.lock.Unlock()
-	psp.sendsymkey = keyid
-	psp.symkeyexpires = time.Now().Add(time.Hour * 24 * 365)
+	self.addSymmetricKeyToPool(keyid, topic, address, expires, addtocache)
 	return keyid, nil
 }
 
-func (self *Pss) deregister(topic *whisper.TopicType, h *Handler) {
+// adds a symmetric key to the pss key pool, and optionally adds the key to the collection of keys used to attempt incoming message decryption
+func (self *Pss) addSymmetricKeyToPool(keyid string, topic whisper.TopicType, address PssAddress, expires time.Duration, addtocache bool) {
 	self.lock.Lock()
 	defer self.lock.Unlock()
-	handlers := self.handlers[*topic]
-	if len(handlers) == 1 {
-		delete(self.handlers, *topic)
-		return
+	var psp *pssPeer
+	if expires == 0 {
+		expires = symKeyExpiry
+	}
+	if _, ok := self.symKeyPool[keyid]; ok == false {
+		self.symKeyPool[keyid] = make(map[whisper.TopicType]*pssPeer)
+		psp = &pssPeer{}
+		self.symKeyPool[keyid][topic] = psp
+	}
+	psp = self.symKeyPool[keyid][topic]
+	psp.expires = time.Now().Add(expires)
+	psp.address = address
+	if addtocache {
```

Comment:
We should adopt the [Go convention](https://golang.org/doc/effective_go.html#mixed-caps) of using `CamelCase` for multiword names (so `addToCache` rather than `addtocache`).

Reply:
I thought this convention is for function names and members? In my view it increases readability to separate *local* vars from these by not using humps.